### PR TITLE
Plot Waterfall Components Decomposition

### PR DESCRIPTION
**Pull request by cetagostini-wise**
## Related Issue

- [ ] Closes #
- [ ] Related to #

## Checklist

- [ ] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [ ] Included tests that prove the fix is effective or that the new feature works
- [ ] Added necessary documentation (docstrings and/or example notebooks)
- [ ] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

## Modules affected

- [ ] MMM
- [ ] CLV

## Type of change

- [ ] New feature / enhancement
- [ ] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):

----
📚 Documentation preview 📚: https://pymc-marketing--625.org.readthedocs.build/en/625/

### CODE REVIEW

File: pymc_marketing/mmm/base.py [hunk 1]
```diff
@@ -1274,6 +1274,124 @@ def plot_channel_contribution_share_hdi(
...
+            dataframe["sin_order_1"]
+            + dataframe["sin_order_2"]
+            + dataframe["cos_order_1"]
+            + dataframe["cos_order_2"]
+        )
+        dataframe.drop(
+            ["sin_order_1", "sin_order_2", "cos_order_1", "cos_order_2"],
+            axis=1,
+            inplace=True,
+        )
```
Comment (juanitorduz) [inline pymc_marketing/mmm/base.py:1315]: We should have a function that extract the seasonal components by looping over "cos_order_*" and "sin_order_*" over all components defined by `yearnly_seasonality` (https://github.com/pymc-labs/pymc-marketing/blob/main/pymc_marketing/mmm/delayed_saturated_mmm.py#L54)

Comment (cetagostini) [reply pymc_marketing/mmm/base.py:1315]: Yes, sounds good. Should we created in this same PR?

Comment (cetagostini) [reply pymc_marketing/mmm/base.py:1315]: I did the following and looks enough:

```python
        if getattr(self, "yearly_seasonality", None):
            contributions_fourier_over_time = pd.DataFrame(
                az.extract(
                    self.fit_result,
                    var_names=["fourier_contributions"],
                    combined=True,
                )
                .mean("sample")
                .to_dataframe()
                .squeeze()
                .unstack()
                .sum(axis=1),
                columns=["yearly_seasonality"],
            )
```

**Comment by ColtAllen** [comment]
I like this plot a lot! Is it possible to flip it around so that the most impactful component is shown first on the X-axis? In visualization theory, our eyes start from the top-left and scan to the right, much like reading a book. If doable, it's best to orient by importance in the same manner.

File: pymc_marketing/mmm/base.py [hunk 2]
```diff
@@ -1274,6 +1276,112 @@ def plot_channel_contribution_share_hdi(
     def graphviz(self, **kwargs):
         return pm.model_to_graphviz(self.model, **kwargs)
 
+    def plot_waterfall_components_decomposition(
+        self, original_scale: bool = True, figsize: Tuple = (14, 7), **kwargs
```
Comment (wd60622) [inline pymc_marketing/mmm/base.py:1280]: can we pass an optional `ax` like in the other plots

File: pymc_marketing/mmm/base.py [hunk 3]
```diff
@@ -1274,6 +1276,112 @@ def plot_channel_contribution_share_hdi(
...
+            The size of the figure. The default is (14, 7).
+        **kwargs
+            Additional keyword arguments to pass to the matplotlib `subplots` function.
+
+        Returns
+        -------
+        fig : matplotlib.figure.Figure
+            The matplotlib figure object.
+        ax : matplotlib.axes.Axes
+            The matplotlib axes object.
```
Comment (wd60622) [inline pymc_marketing/mmm/base.py:1299]: This signature is different than what is being returned

File: pymc_marketing/mmm/base.py (hunk 1 above)
Comment (juanitorduz) [reply pymc_marketing/mmm/base.py]: yeah, you can do it on this PR

File: pymc_marketing/mmm/base.py [hunk 4]
```diff
@@ -1274,6 +1276,112 @@ def plot_channel_contribution_share_hdi(
...
+        stack_dataframe = dataframe.stack().reset_index()
+        stack_dataframe.columns = pd.Index(["date", "component", "contribution"])
+        stack_dataframe.set_index(["date", "component"], inplace=True)
+        dataframe = stack_dataframe.groupby("component").sum()
+        dataframe.sort_values(by="contribution", ascending=True, inplace=True)
+        dataframe.reset_index(inplace=True)
+
+        # Calculate the percentage of each contribution
+        total_contribution = dataframe["contribution"].sum()
+        dataframe["percentage"] = (dataframe["contribution"] / total_contribution) * 100
```
Comment (wd60622) [inline pymc_marketing/mmm/base.py:1316]: I'd suggest to pull this out and test

**Comment by juanitorduz** [comment]
@cetagostini-wise I have seen you have a conflict with the mmm example notebook. Git with notebooks is a nightmare so here is a tip: If you want to rebase and keep the changes in your branch then you run:

```
git rebase --strategy-option theirs main 
```

If you want to keep the changes from `main` so that you can update the notebook yourself, then run

```
git rebase --strategy-option ours main 
```

I hope it helps :)

File: pymc_marketing/mmm/base.py [hunk 5]
```diff
@@ -1146,6 +1146,8 @@ def compute_mean_contributions_over_time(
                 .to_dataframe()
                 .squeeze()
                 .unstack()
+                .sum(axis=1),
+                columns=["yearly_seasonality"],
```
Comment (wd60622) [inline pymc_marketing/mmm/base.py:1150]: How will someone know that `yearly_seasonality` is available in the `plot_grouped_contribution_breakdown_over_time`? 

outside scope of PR: But would it make sense to store this as a deterministic variable in our model?

**Review by wd60622 — COMMENTED** [review]
some questions and comments

File: pymc_marketing/mmm/base.py (hunk 5 above)
Comment (cetagostini-wise) [reply pymc_marketing/mmm/base.py:1150]: > How will someone know that yearly_seasonality is available in the plot_grouped_contribution_breakdown_over_time?

Should be describe in the example.


> outside scope of PR: But would it make sense to store this as a deterministic variable in our model?

I have been thinking on it. I'll make it in another PR

Comment (cetagostini) [reply pymc_marketing/mmm/base.py:1150]: Done!

**Comment by cetagostini** [comment]
> @cetagostini-wise I have seen you have a conflict with the mmm example notebook. Git with notebooks is a nightmare so here is a tip: If you want to rebase and keep the changes in your branch then you run:


As discussed can we do it in another branch?

**Pull request closed without merging**